### PR TITLE
Remove progress bar during snapshot creation

### PR DESCRIPTION
**Pull request by f1ames**
## What is the purpose of this pull request?

Task

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

Progress bar is now removed from HTML during snapshot creation so it does not pollute the images causing undo manager to think something changes.

### CODE REVIEW

**Comment by f1ames** [comment]
As Easy Image was merged to `major` already, I targeted this PR there too.

**Comment by f1ames** [comment]
This PR does not solve all udno integration problems, but only one in which progress bar HTML (which changes due to tracking upload progress), causes undoManager to save new snapshots, because of changes in HTML (caused by progress bar progress).

See more in #1532.

File: tests/plugins/easyimage/manual/progressbarundo.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+## Easy Image Progressbar Undo
+
+1. Upload some images (using drag and drop, copy/paste, etc).
+2. During image upload manipulate selection (click somewhere, select some editor content, etc).
+
+### Expected
+
+**Saved undo steps** counter should not increment on selection manipulation.
+
+**Important**: Selecting or deselecting widget will create an undo step.
```
Comment (Comandeer) [inline tests/plugins/easyimage/manual/progressbarundo.md]: Honestly I don't understand these instructions.

First of all, upload is too quick to allow sensible "selection manipulation". Maybe mocking it with slowed progress would be a good idea?

Comment (Comandeer) [inline tests/plugins/easyimage/manual/progressbarundo.md]: Honestly I don't understand these instructions.

First of all, upload is too quick to allow sensible "selection manipulation". Maybe mocking it with slowed progress would be a good idea?

What's more, how to check if counter does not change, when every selection change will change it?

Comment (f1ames) [reply tests/plugins/easyimage/manual/progressbarundo.md]: Slowing down an upload makes sense, as network speed throttling is only available on Chrome (and probably you will need 10mb image to test it properly without it).

As for selection change adding snapshot, I'm not sure, tested it on Chrome, FF, Safari, IE11 and it seems pure selection changes does not cause new undo step:

feb-08-2018 11-40-48

Maybe there are some particular cases in which id does and some in which it doesn't?

Comment (f1ames) [reply tests/plugins/easyimage/manual/progressbarundo.md]: There is a case with selecting/deselecting widget which triggers undo snapshot, but it is already covered in tests description.

Comment (f1ames) [reply tests/plugins/easyimage/manual/progressbarundo.md]: There are also other cases, like select all. Relaying on undo snapshots count may be misleading. I redesigned the test to check if latest created snapshot contains progressbar.

**Comment by mlewand** [comment]
We won't be able to contain this in the upcomin release. Let's move it to 4.9.1 instead.

File: plugins/imagebase/plugin.js [hunk 2]
```diff
@@ -832,6 +842,44 @@
...
+					// Attach wrapper only if it is detached.
+					parent.append( this.wrapper );
+				}
+			}, this )
+		);
+
+		var remove = this.remove;
+		this.remove = function() {
+			// Remove undo listeners when progress bar is removed.
+			CKEDITOR.tools.array.filter( undoListeners, function( listener ) {
```
Comment (Comandeer) [inline plugins/imagebase/plugin.js]: Why `filter` is used here? It does not modify original array, so it's left untouched, with all listeners inside. `CKEDITOR.tools.array.forEach` + `undoListener.length = 0` seems like a better alternative.

File: plugins/imagebase/plugin.js [hunk 3]
```diff
@@ -832,6 +842,44 @@
...
 
+	ProgressBar.prototype.bindEditor = function( editor ) {
+		var undoListeners = [],
+			parent = this.wrapper.getParent();
+
+		// Hide progressbar during snapshot creation so it does not trigger saving new snapshot
+		// (as content will be changed due to progressbar progress).
+		undoListeners.push(
+			editor.on( 'beforeUndoImage', function() {
+				if ( this.wrapper.getParent() ) {
```
Comment (Comandeer) [inline plugins/imagebase/plugin.js]: This function could be simplified by early return if `this.wrapper` does not have a parent.

File: tests/plugins/imagebase/features/upload.js [hunk 4]
```diff
@@ -608,6 +646,56 @@
...
+					fullLoad: false,
+					callback: function( widgets ) {
+						var initialSnapshots = editor.undoManager.snapshots.length,
+							loader = editor.uploadRepository.loaders[ 0 ];
+
+						assert.areSame( 1, widgets.length, 'Widgets count' );
+
+						loader.uploadTotal = 10;
+
+						updateLoader( loader, 2, function() {
```
Comment (Comandeer) [inline tests/plugins/imagebase/features/upload.js]: This whole part of tests is very repetitve. I'd extract it to dedicated helper function.

File: plugins/imagebase/plugin.js (hunk 2 above)
Comment (f1ames) [reply plugins/imagebase/plugin.js]: Looks like some leftover after poor copy/paste 😭

**Comment by Comandeer** [comment]
I've rebased to latest major and update tests.

**Comment by mlewand** [comment]
Taking over the review.

**Comment by f1ames** [comment]
I have revisited this PR and it looks like some minor thing. Since Easy Image plugin is publicly available for some significant amount of time already and we didn't have any similar issue reports it is a low priority.

As for the solution itself we can continue with the above (it's 1,5 year old so may need some updates) or try to address this with `undoManager.filter` (#3245) mechanism which was recently introduced.

**Pull request closed without merging**